### ISSUE TITLE
Lecture 2. ML Basics

### ISSUE THREAD

**Issue by seonghunn** [seonghunn/CS182#2]
# Lecture 2. ML Basics
### Different types of learning problem
- Supervised learning (지도학습)
- Unsupervised learning (비지도학습)
- Reinforcement learning (강화학습)

### Supervised learning

<img width="1315" alt="image" src="https://user-images.githubusercontent.com/69765278/158071221-39cb6bfd-05c3-4d7f-bb3b-ceb6fcf07c49.png">

- 회귀, 분류에 사용한다.
- label(정답)이 존재하여 x라는 입력이 들어왔을 때 y라는 라벨을 예측하도록 네트워크 parameter를 훈련시킨다.
- 이 네트워크에 해당하는 param인 &theta;를 훈련시킬 때 입력 x와 라벨 y의 pair를 사용한다.
- 통과되어 나온 값 f<sub>&theta;</sub>(x) 가 y와 얼마나 유사한지 판단하기 위하여 loss function을 사용한다.
- 다음으로 이 loss를 줄이기 위하여 gradient descent 등의 방법으로 최적화 한다
- ex) Classic linear regression problem

### Unsupervised learning

<img width="1044" alt="image" src="https://user-images.githubusercontent.com/69765278/158071239-5b996d3f-71bd-4055-80db-1a43424d1fba.png">

- unlabeled data를 이용하여 특징을 추출해 잘 representation 하는 것이 목적이다.
- 특정 이미지를 생성하는 등의 역할을 하기 때문에 label이 필요 없다.
- Generative modeling(GAN, 이미지 생성),  self-supervised representation learning(BERT 모델, 언어쪽에서 representation을 이용)

### Reinforcement learning

<img width="1320" alt="image" src="https://user-images.githubusercontent.com/69765278/158071255-90cd1b3a-3ff5-4021-83ba-28a150b5b580.png">

- agent의 행동이 환경에 영향을 미치고, 그것을 agent가 다시 학습하여 최적의 행동을 할 수 있도록 학습한다
- Game, simulation에 활용 (알파고 등...)
- 지도학습의 일종으로 볼 수 있다 (최적의 행동이 label, 즉 정답이라고 생각할 수 있음)

#### overview
1. 결과를 예측하여 어떤 방법으로 model을 정의할지 생각한다 (regression, classification 등..)
2. 결과의 가정에 기반해 어떤 Loss function을 사용할지 생각한다.  
  ex) regression 모델이 가우시안 분포를 띌 거라고 예상 →MSE loss(continuous한 가우시안)
  ex) classification 모델이 multi -value classification → binary cross entropy
3. optimizer를 선택한다 (대부분 gradient descent를 사용하게 됨)

#### Concept
dataset D를 통해 f<sub>&theta;</sub>(x) = y 가 될 수 있도록 &theta;를 학습시킨다.
p(y|x) 은 x가 주어졌을 때 y가 될 확률이며, 저 **probrability distribution**을 알게 하는 것이다.

<img width="1062" alt="image" src="https://user-images.githubusercontent.com/69765278/158074110-d0c8aa68-182e-40ae-a491-e65349739d6a.png">

하지만 예측은 쉽지 않기 때문에, 확률의 관점에서 학습을 진행한다. 즉 다음과 같이 조건부확률을 도입하게 된다.

<img width="763" alt="image" src="https://user-images.githubusercontent.com/69765278/158074448-cee9fe28-835c-467a-8b8b-7bceeb931e77.png">
<img width="252" alt="image" src="https://user-images.githubusercontent.com/69765278/158074584-bde6fd8b-d116-4dc6-b052-da03a6cc77d5.png">
p(y=dog|x)=x<sup>T</sup>&theta;<sub>dog</sub> 이란 식은 likelihood라고 하는데, 조건부확률의 모습을 띈 x에 대한 함수로 표현할 수 있다. 즉 x가 변수이며 입력 x에 대한 특정 y의 결과가 도출될 확률이란 의미를 가지고 있다. likelihood의 직관적인 의미는 **가정한 분포가 실제 분포와 얼만큼 비슷한가?** 라고 생각할 수 있다.
여기서 softmax 함수를 도입하여 다음과 같은 식을 도입할 수 있다.

<img width="587" alt="image" src="https://user-images.githubusercontent.com/69765278/158074496-91b93183-0e32-4c62-afa0-d617bfd50d17.png">

softmax 함수 이외의 다른
- one to one
- onto
- sum to 1
이 되게 하는 모든 함수를 사용할 수 있는데, 현업과 학계 모두에서 softmax를 많이 사용하고 있다고 한다.

<img width="839" alt="image" src="https://user-images.githubusercontent.com/69765278/158074876-c123eb92-54a5-4116-862a-8b7ac4ec0a01.png">
<img width="700" alt="image" src="https://user-images.githubusercontent.com/69765278/158074918-e94f07b5-ad21-471c-9eb3-7cca931876e1.png">

softmax는 위와 같이 활용된다. 이를 통해 최적의 &theta; 를 학습하게 된다

#### Concept
**모델의 분포와, 입력받은 데이터(훈련 데이터)와의 차이를 나타냄**
#### How is dataset generated?

<img width="698" alt="image" src="https://user-images.githubusercontent.com/69765278/158076908-23a2e1a8-bc35-45d5-908b-71cf4856c9ee.png">

우리가 learning을 하는 과정은 given input에 대하여 해당 라벨에 대한 **Conditional probablity distribution** 을 찾는 것이다.

<img width="853" alt="image" src="https://user-images.githubusercontent.com/69765278/158077096-756f57a4-1e47-45a7-99f4-ebc24faa5593.png">

training set D에 대하여, 각각의 pair 가 전부 독립이고 균일하게 분포되어 있다는 가정하에(i.i.d) D에 대한 확률분포 P(D)는 개별 확률의 곱으로 표현된다.

<img width="883" alt="image" src="https://user-images.githubusercontent.com/69765278/158077213-d5223a0b-7a08-4256-a393-17ae392d62e6.png">

이를 확률의 정의를 써서 조작한 식에서 **likelihood인 p<sub>&theta;</sub>(y|x)의 곱들이 최대가 되도록 만들어 준다면 우리는 D를 통해 결과와 가장 유사한 확률 분포를 나타내는 Parameter &theta; 를 알게된다.** (likelihood가 커질수록 실제 분포에 가까워 짐)
이 식을 조금 더 조작하여 다음과 같은 loss function 들을 도출해 낼 수 있다.

<img width="888" alt="image" src="https://user-images.githubusercontent.com/69765278/158077589-7014bfcc-d67c-4e3c-9060-8570dc48a741.png">
<img width="901" alt="image" src="https://user-images.githubusercontent.com/69765278/158077610-acca07f6-e086-40c4-a601-9f62f9defbd2.png">
(why? 에 대한 답은 supplement에서 언급하겠다)

### Optimization
**대부분 gradient descent를 사용한다**

**Comment by seonghunn** [comment]
# Supplement
[ref](https://github.com/Jasonlee1995/CS182/blob/main/Supplements/Lecture02_sup.pdf)
